### PR TITLE
Correct nested generated column validation during 'Insert' and 'Update'

### PR DESCRIPTION
**Pull request by smakalias**
## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 1]
```diff
@@ -1446,4 +1446,58 @@ public void testInsertDefaultExpressions() {
...
+        execute("insert into t (id, o) values (3, {username='crate', pwd='$$$'})");
+        execute("insert into t (id, o) select 4, {username='crate', pwd='$$$'}");
+
+        execute("refresh table t");
+
+        assertThat(
+            printedTable(execute("select * from t order by id").rows()),
+            is("1| {pwd=***, username=crate}\n" +
+               "2| {pwd=***, username=crate}\n" +
+               "3| {pwd=$$$, username=crate}\n" +
```
Comment (smakalias) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: actual result: the generated value is inserted - no validation

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 2]
```diff
@@ -1446,4 +1446,58 @@ public void testInsertDefaultExpressions() {
...
+               "4| {pwd=$$$, username=crate}\n"
```
Comment (smakalias) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: actual result: the generated value is inserted - no validation

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 3]
```diff
@@ -1446,4 +1446,58 @@ public void testInsertDefaultExpressions() {
...
+
+        // do provide value for nested default expression
+        execute("insert into t (id, o) values (3, {username='crate', pwd='$$$'})");
+        execute("insert into t (id, o) select 4, {username='crate', pwd='$$$'}");
+
+        execute("refresh table t");
+
+        assertThat(
+            printedTable(execute("select * from t order by id").rows()),
+            is("1| {pwd=***, username=crate}\n" +
```
Comment (smakalias) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: actual result: the default value is ignored :/

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 4]
```diff
@@ -1446,4 +1446,58 @@ public void testInsertDefaultExpressions() {
...
+               "2| {pwd=***, username=crate}\n" +
```
Comment (smakalias) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: actual result: the default value is ignored :/

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 5]
```diff
@@ -1427,4 +1427,74 @@ public void testInsertDefaultExpressions() {
...
+        execute("insert into t (id, o) values (3, {username='crate', pwd='***'})");
+        execute("insert into t (id, o) select 4, {username='crate', pwd='***'}");
+
+        // provide 'valid' value for 2nd level only nested generated expression (given == generated)
+        execute("insert into t (id, o) values (5, {username='crate', nested={secret='***'}})");
+        execute("insert into t (id, o) select 6, {username='crate', nested={secret='***'}}");
+
+        // provide 'valid' values for all-level nested generated expressions (given == generated)
+        execute("insert into t (id, o) values (7, {username='crate', pwd='***', nested={secret='***'}})");
+        execute("insert into t (id, o) select 8, {username='crate', pwd='***', nested={secret='***'}}");
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: Why do we need to test for different levels? will this add any value? If not please remove, keep the test scope low, its otherwise hard to read, understand and maintain.

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 6]
```diff
@@ -1427,4 +1427,74 @@ public void testInsertDefaultExpressions() {
...
+            is("1| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "2| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "3| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "4| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "5| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "6| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "7| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "8| {nested={secret=***}, pwd=***, username=crate}\n" +
+               "9| {nested={secret=***}, pwd=***}\n" +
+               "10| {nested={secret=***}, pwd=***}\n"
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: Could we implement unit tests instead? otherwise can we create a test method for each different scenario?

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 7]
```diff
@@ -70,21 +77,55 @@ public GeneratedColumns(InputFactory inputFactory,
...
+    }
+
+    private void processGeneratedExpressionColumn(Reference reference,
+                                                 Input<?> expressionInput,
+                                                 Collection<Reference> presentColumns,
+                                                 Set<String> topLevelPresentColumns,
+                                                 boolean doValidate) {
+        final ColumnIdent ident = reference.column();
+        final boolean isNestedColumn = !ident.isTopLevel();
+        if (isNestedColumn) {
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: I think this is only true on update, can we add a test please?

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 8]
```diff
@@ -52,11 +57,13 @@
 
     private final Map<Reference, Input<?>> toValidate;
     private final Map<Reference, Input<?>> toInject;
+    private final Map<Reference, Input<?>> maybeInjectNested;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: I think we could use clarification here that explains the difference between `toInject` and `maybeInjectNested`. 

The cases in which they apply, etc..

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 9]
```diff
@@ -96,21 +137,53 @@ public void setNextRow(T row) {
...
-                    "Given value " + providedValue +
-                    " for generated column " + target.column() +
-                    " does not match calculation " + ((GeneratedReference) target).formattedGeneratedExpression() + " = " +
-                    generatedValue
-                );
+            doValidateValue(target, input.value(), providedValue);
+        }
+    }
+
+    private void doValidateValue(Reference target, Object generatedValue, Object providedValue) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: Please avoid `x` and `doX` style naming.

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 10]
```diff
@@ -70,21 +77,55 @@ public GeneratedColumns(InputFactory inputFactory,
...
+        maybeInjectNested = new HashMap<>();
+        Set<String> topLevelPresentColumns = presentColumns
+            .stream()
+            .map(Reference::column)
+            .filter(ColumnIdent::isTopLevel)
+            .map(ColumnIdent::name)
+            .collect(Collectors.toSet());
+
+        allGeneratedColumns.forEach(
+            r -> processGeneratedExpressionColumn(
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: Please keep in mind that the `GeneratedColumns` constructions shouldn't become too heavy as this is used for single inserts as well.. 

Lambdas that capture a closure tend to have a bit more bootstrapping overhead than a regular loop.

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 11]
```diff
@@ -96,21 +137,53 @@ public void setNextRow(T row) {
...
+            throw new IllegalArgumentException(
+                "Given value " + providedValue +
+                " for generated column " + target.column() +
+                " does not match calculation " + ((GeneratedReference) target).formattedGeneratedExpression() + " = " +
+                generatedValue
+            );
+        }
+    }
+
+    void checkNestedValues(Reference target, Object providedValue) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: I think this lacks some high level explanation *why* nested references have to be treated in a special why.

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java (hunk 7 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: it can actually happen with a nested generated column, the added tests cover such cases.

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java (hunk 6 above)
Comment (smakalias) [reply sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: Thanks for your comment..

I've added unit tests to cover the specific use cases I wanted to highlight. 
I also removed the long integration test as I saw there were existing tests covering many scenarios, I only added a couple.

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java (hunk 5 above)
Comment (smakalias) [reply sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: Removed, see https://github.com/crate/crate/pull/8811#discussion_r296664485

**Comment by smakalias** [comment]
@mfussenegger I realized that same issue also applies for update statements - this change also covers for that as well - added related tests

Could you please review?

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 12]
```diff
@@ -64,53 +69,92 @@ public GeneratedColumns(InputFactory inputFactory,
...
     }
 
+    private void processGeneratedExpressionColumn(Reference reference,
+                                                  Input<?> expressionInput) {
+        toInject.put(reference, expressionInput);
+        final ColumnIdent ident = reference.column();
+        if (!ident.isTopLevel()) {
+            List<Reference> references = nestedReferenceByTopLevelColumnName.getOrDefault(
+                ident.name(),
+                new ArrayList<>());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: I think a manual `.get()` followed by an `if (.. null)` would be more efficient as right now the `new ArrayList<>()` is always evaluated, not just when there is no element found.

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 13]
```diff
@@ -64,53 +69,92 @@ public GeneratedColumns(InputFactory inputFactory,
...
-        Input<?> input = toValidate.get(target);
+        validateValue(target, providedValue, validation == Validation.VALUE_MATCH);
+    }
+
+    public void validateValue(Reference target, Object providedValue, boolean validate) {
+        if (providedValue == null) {
+            return;
+        }
+
+        Input<?> input = toInject.get(target);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]:
```suggestion
        Input<?> input = toInject.remove(target);
```

If I read the diff correctly currently `toInject.remove` is always called anyway if the key is present. So it is semantically the same.

Maybe add a short note why it is being removed?

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 14]
```diff
@@ -1371,6 +1371,41 @@ public void testInsertWithNestedGeneratedColumn() {
         assertThat(printedTable(response.rows()), is("10| 11| 4\n"));
     }
 
+    @Test
+    public void testInsertWithNestedGeneratedColumnGivenValueMatches() {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: Do we need all these integration level tests? Looks like the unit tests already cover those cases?

Comment (smakalias) [reply sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: It's true that unit tests also cover these, 
but I would keep them for as long as we have two paths (insert into values, insert into select)
to be on the safe side.

What do you think ?

File: sql/src/test/java/io/crate/execution/dml/upsert/UpdateSourceGenTest.java [hunk 15]
```diff
@@ -166,9 +166,9 @@ public void testNestedGeneratedColumnIsGenerated() throws Exception {
     @Test
     public void testNestedGeneratedColumnIsGeneratedValidateValueIfGivenByUser() throws Exception {
         SQLExecutor e = SQLExecutor.builder(clusterService)
-            .addTable("create table t (x long, obj object as (y as x + 1))")
+            .addTable("create table t (x int, obj object as (y as 'foo'))")
```
Comment (smakalias) [inline sql/src/test/java/io/crate/execution/dml/upsert/UpdateSourceGenTest.java]: Not sure why but these 2 tests were failing randomly - when they did fail, the generated expression i.e. `input.value()` resulted in `null`.
Any idea why?

It seems to be a glitch with the test itself, I have verified with an integration test that it is not an issue with the current
changes.

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java (hunk 14 above)
Comment (mfussenegger) [reply sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: They're an order of magnitude or more slower than unit tests, and we don't want the time it takes to run the full test suite to explode over time. 

So I think we should limit the integration tests to "Is the generated columns class integrated correctly" and test all the edge cases and scenarios within the unit tests.

> I would keep them for as long as we have two paths (insert into values, insert into select)
to be on the safe side.

The problem is that we hardly ever notice that tests have become obsolete.

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 16]
```diff
@@ -91,9 +91,10 @@ private void processGeneratedExpressionColumn(Reference reference,
...
         if (!ident.isTopLevel()) {
-            List<Reference> references = nestedReferenceByTopLevelColumnName.getOrDefault(
-                ident.name(),
-                new ArrayList<>());
+            List<Reference> references = nestedReferenceByTopLevelColumnName.get(ident.name());
+            if (references == null) {
+                references = new ArrayList<>();
+            }
             references.add(reference);
             nestedReferenceByTopLevelColumnName.put(ident.name(), references);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: The put could also be moved up into the `== null` block, afterwards the reference that's already inside the dictionary is being mutated, so no additional `put` is necessary.

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java [hunk 17]
```diff
@@ -126,12 +127,13 @@ public void validateValue(Reference target, Object providedValue, boolean valida
             return;
         }
 
-        Input<?> input = toInject.get(target);
+        // remove the reference from the 'toInject' map
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: How about something like:

```
// `toInject` contains columns that might be provided by the user because for nested columns we don't know if the user is going to provide them based on the statement. E.g. in (INSERT INTO (obj) (select ...)) (We know `obj` is going to be provided, but not which nested columns the values will contain.). 
// -> We don't need to inject values that are already provided, so we remove them on the fly.

File: sql/src/test/java/io/crate/execution/dml/upsert/UpdateSourceGenTest.java (hunk 15 above)
Comment (mfussenegger) [reply sql/src/test/java/io/crate/execution/dml/upsert/UpdateSourceGenTest.java]: No clue, but I think we've to investigate this instead of changing the tests.

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java (hunk 17 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java]: E.g. 

```
INSERT INTO t (obj) (SELECT col1 FROM unnest([{a=1, b=2}, {a=2}]);
```

**Comment by mfussenegger** [comment]
Closing this as https://github.com/crate/crate/pull/8873 got merged.
